File: Data Structures & Algorithms/trapping-rain-water/submission-4.java
// class Solution {
//     public int trap(int[] height) {
//         if (height.length == 0) {
//             return 0;
//         }

//         Stack<Integer> stack = new Stack<>();
//         int res = 0;

//         for (int i = 0; i < height.length; i++) {
//             while (!stack.isEmpty() && height[i] >= height[stack.peek()]) {
//                 int mid = height[stack.pop()];
//                 if (!stack.isEmpty()) {
//                     int right = height[i];
//                     int left = height[stack.peek()];
//                     int h = Math.min(right, left) - mid;
//                     int w = i - stack.peek() - 1;
//                     res += h * w;
//                 }
//             }
//             stack.push(i);
//         }
//         return res;
//     }
// }

class Solution {
    public int trap(int[] height) {
        int n = height.length;
        if (n == 0) {
            return 0;
        }

        int left = 0, right = n - 1;
        int lMax = 0, rMax = 0;
        int res = 0;
        while (left < right) {
            lMax = Math.max(lMax, height[left]);
            rMax = Math.max(rMax, height[right]);
            if (lMax < rMax) {
                res += lMax - height[left];
                left++;
            } else {
                res += rMax - height[right];
                right--;
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/generate-parentheses/submission-4.py
# class Solution:
#     def generateParenthesis(self, n: int) -> List[str]:
#         res = []
#         def dfs(s, left, right):
#             if len(s) == 2 * n:
#                 res.append(s)
#                 return
            
#             if left < n:
#                 dfs(s+ "(", left+1, right)
            
#             if right < left:
#                 dfs(s + ")", left, right+1)

#         dfs("", 0,0)
#         return res

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        if n == 0:
            return []
        
        self.res, self.track = [], []
        self.backtrack(n, n)
        return self.res

    def backtrack(self, left, right):
        if right < left:
            return
        
        if left < 0 or right < 0:
            return
        
        if left == 0 and right == 0:
            self.res.append("".join(self.track))
            return
        
        self.track.append("(")
        self.backtrack(left - 1, right)
        self.track.pop()

        self.track.append(")")
        self.backtrack(left, right - 1)
        self.track.pop()

